File: day05/door_hack2.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Digest::MD5 qw(md5_hex);
use 5.24.0;
use feature 'signatures';
no warnings "experimental::signatures";

my $door_id = $ARGV[0];
my $password = ' ' x 8;
my $i = 0;
my $n = 0;

while ($n < 8) {
    my $hash = md5_hex($door_id . $i);
    if (substr($hash, 0, 5) eq '0' x 5) {
        my $pos = substr($hash, 5, 1);
        my $val = substr($hash, 6, 1);
        if ($pos =~ /[0-7]/ && substr($password, $pos, 1) eq ' ') {
            substr($password, $pos, 1) = $val;
            say $password;
            $n++;
        }
    }
    $i++;
}

say $password;

File: day05/door_hack2.py
#!/usr/bin/env python
from sys import argv
import hashlib

script, door_id = argv
password = list('        ')
i = 0
n = 0

while n < 8 :
    k = "%s%i" % (door_id, i)
    m = hashlib.md5(k)
    hash = m.hexdigest()
    if hash[0:5] == "00000" :
        pos = hash[5]
        val = hash[6]
        if pos >= "0" and pos <= "7" and password[int(pos)] == " " :
            password[int(pos)] = val
            print "".join(password)
            n = n + 1
    i = i + 1

print "".join(password)
